spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta
+
+import io.delta.tables._
+
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.catalyst.util.FailFastMode
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.{SharedSparkSession, SQLTestUtils}
+import org.apache.spark.sql.types.StructType
+
+/**
+ * Base trait collecting helper methods to run MERGE tests. Merge test suite will want to mix in
+ * either [[MergeIntoSQLTestUtils]] or [[MergeIntoScalaTestUtils]] to run merge tests using the SQL
+ * or Scala API resp.
+ */
+trait MergeIntoTestUtils extends DeltaDMLTestUtils with MergeHelpers {
+  self: SharedSparkSession =>
+
+  import testImplicits._
+
+  protected def executeMerge(
+      target: String,
+      source: String,
+      condition: String,
+      update: String,
+      insert: String): Unit
+
+  protected def executeMerge(
+      tgt: String,
+      src: String,
+      cond: String,
+      clauses: MergeClause*): Unit
+
+  protected def withCrossJoinEnabled(body: => Unit): Unit = {
+    withSQLConf(SQLConf.CROSS_JOINS_ENABLED.key -> "true") { body }
+  }
+
+  /**
+   * Parse the input JSON data into a dataframe, one row per input element.
+   * Throws an exception on malformed inputs or records that don't comply with the provided schema.
+   */
+  protected def readFromJSON(data: Seq[String], schema: StructType = null): DataFrame = {
+    if (schema != null) {
+      spark.read
+        .schema(schema)
+        .option("mode", FailFastMode.name)
+        .json(data.toDS)
+    } else {
+      spark.read
+        .option("mode", FailFastMode.name)
+        .json(data.toDS)
+    }
+  }
+}
+
+trait MergeIntoSQLTestUtils extends SQLTestUtils with MergeIntoTestUtils {
+  self: SharedSparkSession =>
+
+  protected def basicMergeStmt(
+      target: String,
+      source: String,
+      condition: String,
+      update: String,
+      insert: String): String = {
+    s"""
+       |MERGE INTO $target
+       |USING $source
+       |ON $condition
+       |WHEN MATCHED THEN UPDATE SET $update
+       |WHEN NOT MATCHED THEN INSERT $insert
+      """.stripMargin
+  }
+
+  override def executeMerge(
```

Comment:
Could we add the `protected` access modified for `executeMerge`?

Reply:
Also where we override it, e.g. here

Reply:
Good catch, I though I had it covered the first time